### ISSUE TITLE
Integrate Dimension Insensitive Euclidean Metric (DIEM) into Qdrant

### ISSUE THREAD

**Issue by signalprime** [qdrant/qdrant#4963]
# Feature Request: Integrate Dimension Insensitive Euclidean Metric (DIEM) into Qdrant

## Background
We have come across a paper that proposes a novel distance metric called **Dimension Insensitive Euclidean Metric (DIEM)**, which improves on cosine similarity for multidimensional comparisons, particularly in high-dimensional spaces. Given Qdrant's focus on vector similarity search, implementing DIEM could provide more accurate and interpretable results, especially in scenarios where cosine similarity's dependency on dimensionality might lead to biased outcomes.

### Paper Details
The paper titled "Surpassing Cosine Similarity for Multidimensional Comparisons: Dimension Insensitive Euclidean Metric (DIEM)" provides a comprehensive analysis of the limitations of cosine similarity and introduces DIEM as a more robust alternative. The key advantages of DIEM include:
- Consistent variability across dimensions.
- Elimination of dimensionality-induced biases.
- Better generalizability for high-dimensional data.

**Reference**: 
- [Surpassing Cosine Similarity for Multidimensional Comparisons: Dimension Insensitive Euclidean Metric (DIEM)](https://arxiv.org/abs/2407.08623)

## Objective
We aim to integrate DIEM into Qdrant as an alternative to the existing cosine similarity metric. This will involve:
1. **Adding a DIEM Calculation Function**: Implement the DIEM formula in Rust, which will replace or complement the existing cosine similarity calculation.
2. **Modify Existing Code**: Extend the current vector comparison logic to include DIEM.
3. **Testing and Validation**: Ensure that the implementation is correct and that DIEM performs as expected in various scenarios.

### DIEM Formula
The DIEM is calculated as follows:

```rust
fn diem(a: &[f64], b: &[f64], v_m: f64, v_M: f64, sigma_squared: f64, expected_distance: f64) -> f64 {
    let euclidean_dist = euclidean_distance(a, b);
    let diem_value = (v_M - v_m) / sigma_squared * (euclidean_dist - expected_distance);
    diem_value
}
```

## Key Considerations
- **v_m, v_M, sigma_squared, and expected_distance**: These are parameters that need to be calculated or estimated based on the specific use case and dimensionality of the data.
- Performance Impact: Given the additional calculations involved, we need to ensure that DIEM does not introduce significant performance overhead, particularly in high-dimensional data spaces.

## Request for Collaboration

I do not currently have Qdrant's build environment set up on my system, and I would greatly appreciate help from the team to:

- Implement the DIEM function into Qdrant's Rust codebase.
- Ensure it integrates seamlessly with the existing code.
- Conduct thorough testing to validate the implementation.

## Next Steps
- Assign: Assign this issue to a team member familiar with the codebase and Rust.
- Discussion: Open the floor for any questions or concerns regarding the implementation.
- Implementation: Create a PR that includes the DIEM function and integrates it into the existing vector similarity calculation process.
- Testing: Develop test cases to evaluate the performance and accuracy of DIEM compared to cosine similarity.

Thank you in advance for your collaboration on this feature. I look forward to working together to enhance Qdrant's capabilities!

**Comment by generall** [comment]
Hey @signalprime, do you know any popular embeddings model trained with this metric?

**Comment by signalprime** [comment]
Hi @generall, thanks for the question.  I can't name any.  Would we want to wait for that use case before implementing this?  I may attempt it myself, just need to make a little time

**Comment by generall** [comment]
I can't guarantee that we will merge your PR even if you implement it. As might have seen, distance metric is not only a single formula, we need to maintain multiple implementations for each 

- cpu architecture X quantization type.

Adding a new metric will increase the maintenance cost for us significantly, so before adding one, I would like to see a justification for it.

**Comment by generall** [comment]
Moreover, I see that the proposed distance metric requires some extra parameters. This is also an interface change, which is almost guaranteed to have an impact on the interface

**Comment by signalprime** [comment]
That makes sense.  Yes, when searching for `cosine` I found numerous files and honestly had no idea where to start.  I'd picked `lib/segment/src/spaces/metric_f16/simple_cosine.rs` to look at first but knew that wouldn't be the end of it.  I appreciate your comments, and they shed light on some of that mystery when looking at qdrant as an outsider never having seen the code before yesterday.

**Comment by timvisee** [comment]
A while ago someone implemented manhattan/L1 distance. While it may not be entirely complete, it may give some insight in what might need changing:
- https://github.com/qdrant/qdrant/pull/3079
- https://github.com/qdrant/quantization/pull/21

**Comment by TitasDas** [comment]
Is there evidence that this metric has led to improved outcomes (e.g., more accurate responses) in systems like a Retrieval-Augmented Generation (RAG)? 

If we were to compare it with cosine similarity, what specific methods, metrics, or benchmarks should we be considering in order to build a case for/against the integration of this?

**Comment by lhallee** [comment]
> Is there evidence that this metric has led to improved outcomes (e.g., more accurate responses) in systems like a Retrieval-Augmented Generation (RAG)?
> 
> If we were to compare it with cosine similarity, what specific methods, metrics, or benchmarks should we be considering in order to build a case for/against the integration of this into Qdrant?

Hi all,

I just wanted to add some context to @signalprime's request - I am also looking for implementations of this metric. It is a brand new vector similarity heuristic, but it seems to offer more consistent comparisons compared to Euclidean distance or cosine similarity. For example, here are the histograms of scores using random vectors for Euclidean distance and then DIEM
![image](https://github.com/user-attachments/assets/e727f5b9-0874-4b64-ae35-8f376b3937f8)
![image](https://github.com/user-attachments/assets/6cee0985-f159-478e-b5a2-e757986ee384)
Hence "dimension insensitive"

This has a lot of promise in higher dimensional contrastive learning, as well as using multiple retrieval models concurrently or consistently with different underlying hidden dimensions. It's not currently a lot to motivate actual use or implementation necessarily, but I suspect this will become a dominant metric.

**Comment by lhallee** [comment]
> > For example, here are the histograms of scores using random vectors for Euclidean distance and then DIEM
> 
> Please can you add the details of what exactly you tried here and what you are suggesting through this.
> 
> The paper misses a few details, the most prominent ones being :
> 
> 1. At what range of "high" dimensions does this start being better compared to cosine similarity?
> 2. How does the computational efficiency of this measure up when used with a large set of vectors? (Of course this also depends on the nature of implementation).
> 
> I was trying to arrive at an agreed way of measuring and testing for this so that we can make a good case for adding this to Qdrant (if at all).

Yep, totally agree. I suspect there will be some work done on convergence for certain contrastive tasks with cosine similarity / MNR losses vs DIEM, but it's too new to know any of this stuff. I am not necessarily interested in Qdrant, I was just looking for implementation and came across this thread. Best of luck @signalprime and @TitasDas

**Comment by adistomar** [comment]
> 1. At what range of "high" dimensions does this start being better compared to cosine similarity?

Does anyone know the answer to this question?

**Comment by TitasDas** [comment]
@adistomar Given the silence, I am guessing not. Interested in collaborating to find an answer together to this question?

**Comment by adistomar** [comment]
@TitasDas I don't think I have time to investigate this much. I was actually hoping to find an implementation of DIEM so that I can just test it myself and see if it's gives better results than cosine similarity. But I couldn't find one from the original authors of the paper

**Comment by BodduSriPavan-111** [comment]
> [@TitasDas](https://github.com/TitasDas) I don't think I have time to investigate this much. I was actually hoping to find an implementation of DIEM so that I can just test it myself and see if it's gives better results than cosine similarity. But I couldn't find one from the original authors of the paper

@timvisee @generall @TitasDas @signalprime @adistomar  we have recently published a Python library **"diemsim"**, which implements "Diemsion Insensitive Euclidean Metric" by utilizing NumPy and vectorization concepts. Hope, it will help in this regard.

Feel free to explore the repository here: https://github.com/BodduSriPavan-111/diemsim
